src/coreclr/jit/morph.cpp

```diff
@@ -11328,15 +11318,104 @@ GenTree* Compiler::fgOptimizeBitwiseAnd(GenTreeOp* andOp)
     GenTree* op1 = andOp->gtGetOp1();
     GenTree* op2 = andOp->gtGetOp2();
 
-    // Fold "cmp & 1" to just "cmp".
-    if (andOp->TypeIs(TYP_INT) && op1->OperIsCompare() && op2->IsIntegralConst(1))
+    // X & X -> X. Intentionally enabled only for cheap trees for better throughput.
+    if ((op2->IsLocal() || op2->IsInvariant()) && GenTree::Compare(op1->gtEffectiveVal(), op2))
     {
-        DEBUG_DESTROY_NODE(op2);
-        DEBUG_DESTROY_NODE(andOp);
+        return op1;
+    }
+
+    // X & 0 -> 0
+    if (op2->IsIntegralConst(0))
```

Comment:
Why do this here instead of in `valuenum` as is done for `GT_MUL`, `GT_ADD`, and a few others (https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/valuenum.cpp#L4709)?

Reply:
Seemingly some of them should also be handled in `gtFoldExprConst`, as the basic arithmetic operations look to be?

Reply:
Looks like maybe the basic arithmetic are handled in all three places (morph, valuenum, and gtFoldExpr)

Reply:
> Why do this here instead of in `valuenum` as is done for `GT_MUL`, `GT_ADD`, and a few others (https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/valuenum.cpp#L4709)?

this specific one is already there, I think it's ok to have it in morph too (actually, I should probably move this to gtFoldExp to catch importer as well).

E.g. "X & X -> X" and "X | X -> X" can't be currently expressed in VN - it will require additional work in value prop I believe.

Reply:
> Looks like maybe the basic arithmetic are handled in all three places (morph, valuenum, and gtFoldExpr)

It's ok, an opt can be expressed up to 4 times: Importer, Morph, VN and Lower. And we can only share the impl between Importer and Morph.

Reply:
> E.g. "X & X -> X" and "X | X -> X" can't be currently expressed in VN - it will require additional work in value prop I believe.

It actually looks like there is actually already a path there: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/valuenum.cpp#L4952-L4956

(same for `x ^ x` and `x | x`), is this not actually working today or is it subtly different somehow?

Reply:
> (same for x ^ x and x | x), is this not actually working today or is it subtly different somehow?

It may work for CSE, but we won't replace X | X with X using that VN, we currently only can replace trees with constants, we can't simplify them.

Reply:
Ah, gotcha. That's a bit unfortunate as it mean's we're tracking a few different opts in valuenum, but not actually utilizing them.

Thanks for answering the questions, things make sense to me now 😄

Reply:
> Ah, gotcha. That's a bit unfortunate as it mean's we're tracking a few different opts in valuenum, but not actually utilizing them.

Yeah, I believe Andy raised that point a few times - that we should be able to simplify trees based on VNs. Example:

`(((X+1) + 1) + 1))` can be numbered as `VNF_GT_ADD(X, 3)` and then value prop phase materializes it to actual `X + 3` 🙂 

Actually, this PR is just a quick draft, I think I'll remove these which won't have diffs